File: c++/318. Maximum Product of Word Lengths.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) {
        int lw = words.size();
        vector<pair<int, int>> v;
        for (auto &w : words)
            v.push_back(make_pair(get_mask(w), w.size()));
        
        int res = 0;
        for (int i = 0; i < lw; i++) {
            for (int j = i + 1; j < lw; j++) {
                if ((v[i].first & v[j].first) == 0)
                    res = max(res, v[i].second*v[j].second);
            }
        }
        
        return res;
    }
    
    int get_mask(string &w) {
        int res = 0;
        for (auto &ch : w) 
            res |= 1<<(ch-'a');
        return res;
    }
};

File: python/318. Maximum Product of Word Lengths.py
class Solution(object):
    def maxProduct(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        def get_mask(word):
            i = 0
            for w in word:
                i |= 1<<(ord(w)-ord('a'))
            return i
        
        res = 0
        n = len(words)
        for i in range(n):
            for j in range(i+1, n):
                if get_mask(words[i]) & get_mask(words[j]) != 0:
                    continue
                res = max(res, len(words[i]) * len(words[j]))  

        return res       

if __name__ == '__main__':
    words = ["abcw","baz","foo","bar","xtfn","abcdef"]
    su = Solution()
    print su.maxProduct(words)        

0b010000000000000000000111 
0b100010000010000000100000
